GameOfRunes/GameOfRunes/GameEngine/GameEngine.swift

```diff
@@ -26,30 +29,61 @@ class GameEngine {
         self.removeDelegate = RemoveDelegate(gameEngine: self)
         self.gameStateMachine = gameStateMachine
         self.gameStateMachine?.gameEngine = self
+        
+        EntityType.allCases.forEach { entityType in
+            entities[entityType] = Set()
+        }
     }
     
     func add(_ entity: GKEntity) {
-        guard entities.insert(entity).inserted else {
-            return
+        var isInserted: Bool?
+        
+        switch entity {
```

Comment:
Perhaps extend GKEntity to conform to a protocol that has entity type in it? Then dun need to repeatedly use switch statements.

Reply:
Okay wait, maybe need to make custom entities class conform instead.

Reply:
We can't do that, mainly because we can't override methods in an extension so we would not be able to do polymorphism for a `getType()` method and at the same time using it as a `GKEntity`

Reply:
What about an intermediate class between custom entities and GKEntity?

Reply:
Alternatively,
<img width="277" alt="Screenshot 2020-03-13 at 11 34 34 PM" src="https://user-images.githubusercontent.com/36080705/76635990-415daa00-6583-11ea-87c4-af7039593ca8.png">
<img width="269" alt="Screenshot 2020-03-13 at 11 34 41 PM" src="https://user-images.githubusercontent.com/36080705/76635999-43c00400-6583-11ea-869c-7bf8916f8f00.png">

So the GKEntity will have some default type.